hail/src/main/scala/is/hail/variant/ReferenceGenome.scala

```diff
@@ -343,33 +344,41 @@ case class ReferenceGenome(name: String, contigs: Array[String], lengths: Map[St
 
     val fastaPath = fs.fileStatus(fastaFile).getPath.toString
     val indexPath = fs.fileStatus(indexFile).getPath.toString
-    fastaReader = FASTAReader(ctx.localTmpdir, fs, this, fastaPath, indexPath)
+    fastaReaderCfg = FASTAReaderConfig(ctx.localTmpdir, fs.broadcast, this, fastaPath, indexPath)
   }
 
   def addSequenceFromReader(tmpdir: String, fs: FS, fastaFile: String, indexFile: String, blockSize: Int, capacity: Int): ReferenceGenome = {
-    fastaReader = new FASTAReader(tmpdir, fs.broadcast, this, fastaFile, indexFile, blockSize, capacity)
+    fastaReaderCfg = FASTAReaderConfig(tmpdir, fs.broadcast, this, fastaFile, indexFile, blockSize, capacity)
     this
   }
 
-  def getSequence(contig: String, position: Int, before: Int = 0, after: Int = 0): String = {
+  private val realFastaReader: ThreadLocal[FASTAReader] = new ThreadLocal[FASTAReader]
```

Comment:
should be transient and with a default value of `null`, I think?

Reply:
will need 2 levels of null handling in the fastaReader method I think

Reply:
I've tested this, and it works, in a cluster even. There's no need for the extra layer of transience/null checks.

Reply:
have you tested with a pipeline that would initialize the fasta reader on the driver machine before generating a spark stage?

I'm not sure how to reason about what's serialized if the value is defined on the driver.

perhaps:

```
ht = ht.annotate_globals(foo = hl.locus('1', 100000).sequence_context())
ht = ht.annotate_globals(foo_glob = ht.foo, foo_row = ht.locus.sequence_context())
```

Reply:
Just did. Worked fine.

Reply:
Not sure how this is working though, since ThreadLocal is not serializable in java, but none of this seems to effect scala's serialization.

Reply:
I'm pretty sure that making this `@transient lazy val` is close to what we should do anyways, so I just did that.